### PR TITLE
Multiple branches support

### PR DESCRIPTION
**Pull request by paulovmr**
AF-1269: Index all branches of a project
AF-1270: Allow the selection of different branches
AF-1271: Allow the creation of a branch
AF-1272: Build/deploy actions should be applied to the currently selected branch
AF-1273: Integrate library branch with project explorer branch

Part of an ensemble:
* https://github.com/kiegroup/appformer/pull/437
* https://github.com/kiegroup/kie-wb-common/pull/2046
* https://github.com/kiegroup/drools-wb/pull/913

### CODE REVIEW

File: kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java [hunk 1]
```diff
@@ -45,6 +47,8 @@
 public class ValidationServiceImpl
         implements ValidationService {
 
+    private final Pattern branchNameValidator = Pattern.compile("^(?!/|\\.|.*([/.]\\.|//|@|\\\\))[^\000-\037\177 ~^:?*\\[/]+(?<!\\.lock|[\\.])$");
```
Comment (jhrcek) [inline kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java]: Magic :open_mouth:  Where did you get this from :laughing: Nice test coverage though..

**Comment by jhrcek** [comment]
Found a serious issue during local testing: [RHPAM-1476](https://issues.jboss.org/browse/RHPAM-1476)

File: kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java (hunk 1 above)
Comment (paulovmr) [reply kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java]: I've found a similar one in stack overflow and adapted it to our needs. :)

**Comment by paulovmr** [comment]
@jhrcek The "File System closed" issue should be fixed now. 

About [RHPAM-1476](https://issues.jboss.org/browse/RHPAM-1476), this behavior is expected if the user does not change the pom.xml (Project Settings) GAV (usually the version in this case).

**Comment by jhrcek** [comment]
@paulovmr thanks for the fix, I can confirm that most of the occurences of "File System is closed" are gone. However there is still one left, that I can reproduce manually like this:
1. Create space "s"
2. Open it and create project "p"
3. Delete the newly created project. You'll see the "unexpected error" modal and the following appears in server.log

<details>
  <summary>stacktrace</summary>
  <pre>
13:24:53,431 ERROR [org.guvnor.structure.backend.repositories.RepositoryServiceImpl] (default task-63) Error during remove repository: java.lang.IllegalStateException: FileSystem is closed.
	at org.uberfire.java.nio.fs.jgit.JGitFileSystemImpl.checkClosed(JGitFileSystemImpl.java:321)
	at org.uberfire.java.nio.fs.jgit.JGitFileSystemImpl.getRootDirectories(JGitFileSystemImpl.java:165)
	at org.uberfire.ext.metadata.io.IOServiceIndexedImpl.cleanupDeletedFS(IOServiceIndexedImpl.java:441)
	at org.uberfire.ext.metadata.io.IOServiceIndexedImpl.delete(IOServiceIndexedImpl.java:432)
	at org.guvnor.structure.backend.repositories.RepositoryServiceImpl.doRemoveRepository(RepositoryServiceImpl.java:402)
	at org.guvnor.structure.backend.repositories.RepositoryServiceImpl.removeRepository(RepositoryServiceImpl.java:348)
	at org.guvnor.structure.backend.repositories.RepositoryServiceImpl$Proxy$_$$_WeldClientProxy.removeRepository(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.errai.bus.server.io.AbstractRPCMethodCallback.invokeMethodFromMessage(AbstractRPCMethodCallback.java:65)
	at org.jboss.errai.bus.server.io.VoidRPCEndpointCallback.callback(VoidRPCEndpointCallback.java:36)
	at org.jboss.errai.bus.server.io.RemoteServiceCallback.callback(RemoteServiceCallback.java:54)
	at org.jboss.errai.cdi.server.CDIExtensionPoints$2.callback(CDIExtensionPoints.java:448)
	at org.jboss.errai.bus.server.DeliveryPlan.deliver(DeliveryPlan.java:47)
	at org.jboss.errai.bus.server.ServerMessageBusImpl.sendGlobal(ServerMessageBusImpl.java:297)
	at org.jboss.errai.bus.server.SimpleDispatcher.dispatchGlobal(SimpleDispatcher.java:46)
	at org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:96)
	at org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:113)
	at org.jboss.errai.bus.server.servlet.DefaultBlockingServlet.doPost(DefaultBlockingServlet.java:144)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.undertow.websockets.jsr.JsrWebSocketFilter.doFilter(JsrWebSocketFilter.java:130)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.uberfire.ext.security.server.SecureHeadersFilter.doFilter(SecureHeadersFilter.java:110)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.uberfire.ext.security.server.SecurityIntegrationFilter.doFilter(SecurityIntegrationFilter.java:70)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:65)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.DisableCacheHandler.handleRequest(DisableCacheHandler.java:33)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1501)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1501)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1501)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1501)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:330)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:812)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
  </pre>
</details>

**Comment by jhrcek** [comment]
One more issue filed in jira: [AF-1503](https://issues.jboss.org/browse/AF-1503)

**Comment by paulovmr** [comment]
jenkins retest this

**Comment by paulovmr** [comment]
@jhrcek I've fixed the "File System is closed" left, could you try it again? 

As we talked about today, the [AF-1503](https://issues.jboss.org/browse/AF-1503) issue is happening on master as well, so could be fixed out of this PR's scope.

**Comment by jhrcek** [comment]
@paulovmr looks good, I didn't encounter any more ""File System is closed" exceptions. Update the description of AF-1503 so it's clear it happens on master too.

**Comment by domhanak** [comment]
jenkins retest this

**Comment by barboras7** [comment]
Hi @paulovmr, 
I was trying to test this PR locally on IE. However, when I try to create a project, I get this error: 
screenshot from 2018-08-27 13-33-33
It seems to be a browser specific issue as I am able to create the project on Firefox running on Windows, whereas it does not work on IE. I am using IE11 and Windows 8.1. On what version of Internet Explorer did you try this?

File: kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/main/java/org/kie/workbench/common/screens/library/client/util/breadcrumb/ProjectBranchBreadcrumb.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    public void newBranchEvent(@Observes final NewBranchEvent newBranchEvent) {
+        final Repository repository = newBranchEvent.getRepository();
+        final Space space = repository.getSpace();
+        final String repositoryAlias = repository.getAlias();
+
+        final Space activeSpace = libraryPlaces.getActiveSpace().getSpace();
+        final Repository activeRepository = libraryPlaces.getActiveWorkspace().getRepository();
+        final String activeRepositoryAlias = activeRepository.getAlias();
+
+        if (space.equals(activeSpace) && repositoryAlias.equals(activeRepositoryAlias) && sessionInfo.getIdentity().equals(newBranchEvent.getUser())) {
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/main/java/org/kie/workbench/common/screens/library/client/util/breadcrumb/ProjectBranchBreadcrumb.java]: @paulovmr I think that I already seen this code more then once, maybe we should write a utility method/context object that given a repository see if it's active (or even delegate this for livraby places)

File: kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/main/java/org/kie/workbench/common/screens/library/client/util/breadcrumb/ProjectBranchBreadcrumbView.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public void init(final ProjectBranchBreadcrumb presenter) {
+        final Branch currentBranch = presenter.getCurrentBranch();
+
+        final List<Item<Branch>> items = presenter.getBranches().stream()
+                .map(b -> new Item<>(b, b.getName(), b.equals(currentBranch) ? CHECKED : UNCHECKED, presenter::onBranchChanged, Item.Type.OPTION))
+                .collect(toList());
+
+        items.add(new Item<>(null, "", UNCHECKED, null, Item.Type.SEPARATOR));
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/main/java/org/kie/workbench/common/screens/library/client/util/breadcrumb/ProjectBranchBreadcrumbView.java]: Can you please don't pass null as arguments?

**Comment by paulovmr** [comment]
jenkins retest this

**Comment by paulovmr** [comment]
@barboras7 @ederign I was able to find and, after some testing, verified locally that the issue with IE11 is fixed! \o/

@barboras7 If you could verify it tomorrow, we can merge the PRs.

@ederign I will incorporate your suggested changes in the next PRs.

**Comment by barboras7** [comment]
Great job, @paulovmr! I tested the new version on Internet Explorer and it worked as expected.:blush:

I found two small UI issues, see their description below. However, I believe that these are not serious issues, I'll create Jiras for them and this PR can be merged.

1. The dropdown with branches is a bit broken (the dropdown expands in both directions - up and down, whereas the expected behaviour would be to expand downwards)
screenshot from 2018-08-29 12-17-22
2. The arrows at both ends of the scroll bar do not fit into the corners and the line with "Add Branch" is not centered correctly. 
screenshot from 2018-08-29 12-39-27

**Pull request merged**